File: backend/src/main/java/kr/co/cudo/authoring/common/util/MaskRleConverter.java
package kr.co.cudo.authoring.common.util;

/**
 * CVAT MASK ↔ RLE 변환 자바 포팅 (portable-modules/02).
 *
 * 원본:
 *  - cvat/apps/dataset_manager/formats/transformations.py:104-120 (rle())
 *  - cvat-canvas/src/typescript/shared.ts:408-426 (imageDataToRLE)
 *
 * CVAT 자체 RLE 형식:
 *  - 첫 run 은 항상 0(off) 픽셀 카운트.
 *  - 첫 픽셀이 on 이면 맨 앞에 길이 0 인 off-run 삽입.
 *  - 형식: [run0, run1, run2, ..., runN]   (BBOX 4개 값은 호출부에서 별도 부착)
 *
 * 보안:
 *  - CWE-770 DoS 방어: 픽셀 총 수 1,000,000 (1M) 제한.
 *    SAM2 / Brush 결과의 평균 mask 크기는 수만 픽셀 → 1M 면 충분히 큼.
 */
public final class MaskRleConverter {

    /** CWE-770 DoS 방어 — 단일 mask 최대 픽셀 수. */
    public static final int MAX_PIXELS = 1_000_000;

    private MaskRleConverter() {}

    /**
     * 2D mask → CVAT RLE.
     * mask[y][x] (height × width).
     */
    public static int[] maskToRle(boolean[][] mask) {
        if (mask == null || mask.length == 0) {
            return new int[0];
        }
        int h = mask.length;
        int w = mask[0].length;
        long total = (long) h * w;
        if (total > MAX_PIXELS) {
            throw new IllegalArgumentException("MASK 크기가 최대 허용치 초과 ("
                    + MAX_PIXELS + " 픽셀): " + total);
        }

        // flatten row-major
        boolean[] flat = new boolean[(int) total];
        for (int y = 0; y < h; y++) {
            if (mask[y].length != w) {
                throw new IllegalArgumentException("MASK 가 직사각형이 아닙니다: row " + y + " width=" + mask[y].length);
            }
            System.arraycopy(mask[y], 0, flat, y * w, w);
        }

        // CVAT 규칙: 첫 픽셀이 on 이면 맨 앞에 0-length off-run 삽입.
        java.util.ArrayList<Integer> rle = new java.util.ArrayList<>();
        if (flat[0]) {
            rle.add(0);
        }
        boolean cur = false; // 시작 가정 (off)
        if (flat[0]) {
            cur = true; // 첫 0-run 다음은 on
        }
        int count = 0;
        for (int i = 0; i < flat.length; i++) {
            boolean v = flat[i];
            if (v == cur) {
                count++;
            } else {
                rle.add(count);
                cur = v;
                count = 1;
            }
        }
        rle.add(count);

        int[] result = new int[rle.size()];
        for (int i = 0; i < rle.size(); i++) {
            result[i] = rle.get(i);
        }
        return result;
    }

    /**
     * CVAT RLE → 2D mask.
     * width / height 는 호출부에서 BBOX 로부터 산출하여 전달.
     */
    public static boolean[][] rleToMask(int[] rle, int width, int height) {
        if (rle == null) {
            throw new IllegalArgumentException("RLE 가 null 입니다.");
        }
        long total = (long) width * height;
        if (total > MAX_PIXELS) {
            throw new IllegalArgumentException("MASK 크기가 최대 허용치 초과 ("
                    + MAX_PIXELS + " 픽셀): " + total);
        }
        if (width <= 0 || height <= 0) {
            throw new IllegalArgumentException("width/height 는 양수여야 합니다: " + width + "x" + height);
        }

        boolean[][] mask = new boolean[height][width];
        boolean cur = false; // CVAT 규칙: 항상 off 에서 시작
        int idx = 0;
        for (int run : rle) {
            for (int k = 0; k < run; k++) {
                if (idx >= total) {
                    throw new IllegalArgumentException("RLE 길이 합이 " + total + " 초과");
                }
                int y = idx / width;
                int x = idx % width;
                mask[y][x] = cur;
                idx++;
            }
            cur = !cur;
        }
        return mask;
    }
}

File: frontend/src/features/label/canvas/utils/maskRleConverter.ts
// CVAT MASK ↔ RLE 변환 (TS 포팅).
// 참고: docs/analysis/portable-modules/02-mask-rle-conversion.md
// BE Java MaskRleConverter와 라운드트립 동등성 보장.
//
// CVAT RLE 형식: [run0, run1, ..., runN, left, top, right, bottom]
// - 첫 run은 항상 0(off) 픽셀 개수로 시작
// - 마지막 4개는 inclusive BBOX (right/bottom 포함)
//
// 보안: 1M 픽셀 한도 (frontend-performance + DoS 방어).

export interface BBox {
  left: number;
  top: number;
  right: number;
  bottom: number;
}

const MAX_PIXELS = 1_000_000;

/**
 * 2D boolean mask + bbox → CVAT RLE.
 * mask는 bbox로 cropped된 tight binary mask. 행=top..bottom, 열=left..right.
 */
export function maskToRle(mask: boolean[][], bbox: BBox): number[] {
  const h = bbox.bottom - bbox.top + 1;
  const w = bbox.right - bbox.left + 1;
  if (h * w > MAX_PIXELS) {
    throw new Error(`Mask 픽셀 한도 초과: ${h * w} > ${MAX_PIXELS} (too large)`);
  }
  // row-major flatten
  const flat: number[] = new Array(h * w);
  for (let y = 0; y < h; y++) {
    const row = mask[y] ?? [];
    for (let x = 0; x < w; x++) {
      flat[y * w + x] = row[x] ? 1 : 0;
    }
  }
  return [...rleEncode(flat), bbox.left, bbox.top, bbox.right, bbox.bottom];
}

/**
 * CVAT RLE → tight 2D boolean mask + bbox.
 */
export function rleToMask(rle: number[]): { mask: boolean[][]; bbox: BBox } {
  if (rle.length < 4) {
    throw new Error('Invalid RLE: 길이 부족 (bbox 누락)');
  }
  const left = rle[rle.length - 4];
  const top = rle[rle.length - 3];
  const right = rle[rle.length - 2];
  const bottom = rle[rle.length - 1];
  const h = bottom - top + 1;
  const w = right - left + 1;
  const runs = rle.slice(0, -4);
  const flat = rleDecode(runs, h * w);
  const mask: boolean[][] = [];
  for (let y = 0; y < h; y++) {
    const row: boolean[] = [];
    for (let x = 0; x < w; x++) {
      row.push(flat[y * w + x] === 1);
    }
    mask.push(row);
  }
  return { mask, bbox: { left, top, right, bottom } };
}

/**
 * RGBA Uint8ClampedArray (alpha 채널) → CVAT raw RLE (BBox 미포함).
 * 첫 run은 항상 0 픽셀 (CVAT 규칙) — alpha[0]가 1이면 [0, ...] 삽입.
 */
export function imageDataToRLE(
  imageData: Uint8ClampedArray,
  width: number,
  height: number,
): number[] {
  const totalPixels = width * height;
  if (totalPixels > MAX_PIXELS) {
    throw new Error(`ImageData 픽셀 한도 초과: ${totalPixels} > ${MAX_PIXELS} (too large)`);
  }
  if (imageData.length < totalPixels * 4) {
    throw new Error('ImageData 길이가 width*height*4보다 작습니다');
  }
  const rle: number[] = [];
  let prev = 0;
  let summ = 0;
  // CVAT TS 원본 (shared.ts:408-426) 동일 — alpha 채널만 본다
  for (let i = 3; i < totalPixels * 4; i += 4) {
    const alpha = imageData[i] > 0 ? 1 : 0;
    if (prev !== alpha) {
      rle.push(summ);
      prev = alpha;
      summ = 1;
    } else {
      summ++;
    }
  }
  rle.push(summ);
  return rle;
}

/**
 * CVAT raw RLE → RGBA Uint8ClampedArray.
 * on 픽셀은 흰색 alpha=255, off는 alpha=0.
 */
export function rleToImageData(
  rle: number[],
  width: number,
  height: number,
): Uint8ClampedArray {
  const totalPixels = width * height;
  if (totalPixels > MAX_PIXELS) {
    throw new Error(`Width*Height 한도 초과: ${totalPixels} > ${MAX_PIXELS} (too large)`);
  }
  const out = new Uint8ClampedArray(totalPixels * 4);
  let idx = 0;
  let val = 0; // CVAT 규칙: 첫 run은 off
  for (const run of rle) {
    for (let k = 0; k < run; k++) {
      const offset = (idx + k) * 4;
      if (val === 1) {
        out[offset] = 255;
        out[offset + 1] = 255;
        out[offset + 2] = 255;
        out[offset + 3] = 255;
      }
    }
    idx += run;
    val = 1 - val;
  }
  return out;
}

/**
 * 빈 RLE 체크 (masksHandler.ts:641 패턴).
 */
export function isEmptyRle(rle: number[]): boolean {
  return rle.length < 6;
}

// 내부 헬퍼: 0/1 flat 배열 → run-length 시퀀스 (CVAT 규칙)
function rleEncode(flat: number[]): number[] {
  const n = flat.length;
  if (n === 0) return [];
  const runs: number[] = [];
  let prev = 0; // 항상 0(off)에서 시작 가정
  let summ = 0;
  for (let i = 0; i < n; i++) {
    const v = flat[i] ? 1 : 0;
    if (v !== prev) {
      runs.push(summ);
      prev = v;
      summ = 1;
    } else {
      summ++;
    }
  }
  runs.push(summ);
  return runs;
}

// 내부 헬퍼: run-length 시퀀스 → 0/1 flat 배열
function rleDecode(runs: number[], totalSize: number): number[] {
  const out = new Array<number>(totalSize).fill(0);
  let idx = 0;
  let val = 0; // 첫 run은 off
  for (const run of runs) {
    for (let k = 0; k < run; k++) {
      if (idx + k < totalSize) out[idx + k] = val;
    }
    idx += run;
    val = 1 - val;
  }
  return out;
}
